python/private/py_wheel.bzl

```diff
@@ -203,6 +203,427 @@ _DESCRIPTION_FILE_EXTENSION_TO_TYPE = {
+        return lambda token: token != reference
+
+    def in_(reference):
+        """Predicate testing if a token is in the list `reference`."""
+        return lambda token: token in reference
+
+    def accept(predicate, value):
+        """If `predicate` matches the next token, accept the token.
+
+        Accepting the token means adding it (according to `value`) to
+        the running results maintained in context["norm"] and
+        advancing the cursor in context["start"] to the next token in
+        `version`.
+
+        Args:
+          predicate: function taking a token and returning a boolean
+            saying if we want to accept the token.
+          value: the string to add if there's a match, or, if `value`
+            is a function, the function to apply to the current token
+            to get the string to add.
+
+        Returns:
+          whether a token was accepted.
+        """
+
+        context = contexts[-1]
+
+        if context["start"] >= len(version):
+            return False
+
+        token = version[context["start"]]
+
+        if predicate(token):
+            if type(value) in ["function", "builtin_function_or_method"]:
+                value = value(token)
+
+            context["norm"] += value
+            context["start"] += 1
+            return True
+
+        return False
+
+    def accept_placeholder():
+        """Accept a Bazel placeholder.
+
+        Placeholders aren't actually part of PEP 440, but are used for
+        stamping purposes. A placeholder might be
+        ``{BUILD_TIMESTAMP}``, for instance. We'll accept these as
+        they are, assuming they will expand to something that makes
+        sense where they appear. Before the stamping has happened, a
+        resulting wheel file name containing a placeholder will not
+        actually be valid.
+
+        """
+        context = open_context(contexts[-1]["start"])
+
+        if not accept(is_("{"), str):
+            contexts.pop()
+            return False
+
+        start = context["start"]
+        for _ in range(start, len(version) + 1):
+            if not accept(is_not("}"), str):
+                break
+
+        if not accept(is_("}"), str):
+            contexts.pop()
+            return False
+
+        close_context()
+        return True
+
+    def accept_digits():
+        """Accept multiple digits (or placeholders)."""
+
+        def isdigit(token):
+            return token.isdigit()
+
+        context = open_context(contexts[-1]["start"])
+        start = context["start"]
+
+        for i in range(start, len(version) + 1):
+            if not accept(isdigit, str) and not accept_placeholder():
+                if i - start >= 1:
+                    if context["norm"].isdigit():
+                        # PEP 440: Integer Normalization
+                        context["norm"] = str(int(context["norm"]))
+                    close_context()
+                    return True
```

Comment:
Consider making a `parser.accept()` method that returns True. This will simplify the code quite a bit.